File: automatepro_cpp_tutorials/src/io/digital_out.cpp
#include <chrono>
#include <memory>
#include <mutex>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <automatepro_interfaces/msg/digital_out.hpp>

class DigitalOutPublisher : public rclcpp::Node
{
public:
  DigitalOutPublisher()
  : Node("digital_out_publisher"),
    duty_cycle_sequence_{0, 50, 100, 50},
    sequence_index_(0)
  {
    publisher_ = this->create_publisher<automatepro_interfaces::msg::DigitalOut>(
      "/io/digital_out",
      10);
    timer_ = this->create_wall_timer(
      std::chrono::seconds(1),
      std::bind(&DigitalOutPublisher::timer_callback, this));
  }

  void switch_off()
  {
    std::lock_guard<std::mutex> lock(mutex_);
    stopped_ = true;
    timer_->cancel();
    auto msg = automatepro_interfaces::msg::DigitalOut();
    msg.d_out_pin_id = automatepro_interfaces::msg::DigitalOut::DIGITAL_OUT_H_01;
    msg.duty_cycle_percent = 0;
    publisher_->publish(msg);
    // Fast DDS acknowledges on the writer heartbeat, sent every 3 s by default.
    if (publisher_->wait_for_all_acked(std::chrono::seconds(4))) {
      RCLCPP_INFO(this->get_logger(), "Switched DIGITAL_OUT_H_01 off");
    } else {
      RCLCPP_WARN(this->get_logger(), "DIGITAL_OUT_H_01 off command not acknowledged");
    }
  }

private:
  void timer_callback()
  {
    std::lock_guard<std::mutex> lock(mutex_);
    if (stopped_) {
      return;
    }
    auto msg = automatepro_interfaces::msg::DigitalOut();
    msg.d_out_pin_id = automatepro_interfaces::msg::DigitalOut::DIGITAL_OUT_H_01;     // Digital Out Pin 01
    msg.duty_cycle_percent = duty_cycle_sequence_[sequence_index_];     // Duty Cycle: 0%, 50%, 100%, 50%
                                                                        // 0% - OFF, 100% - ON
    publisher_->publish(msg);
    RCLCPP_INFO(
      this->get_logger(), "Publishing: d_out_pin_id=%d, duty_cycle_percent=%d", msg.d_out_pin_id,
      msg.duty_cycle_percent);

    // Update the sequence index
    sequence_index_ = (sequence_index_ + 1) % duty_cycle_sequence_.size();
  }

  rclcpp::Publisher<automatepro_interfaces::msg::DigitalOut>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::vector<int> duty_cycle_sequence_;
  size_t sequence_index_;
  std::mutex mutex_;
  bool stopped_{false};
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<DigitalOutPublisher>();
  // The IO controller keeps the last command, so switch the output off before shutdown.
  std::weak_ptr<DigitalOutPublisher> weak_node = node;
  rclcpp::contexts::get_global_default_context()->add_pre_shutdown_callback(
    [weak_node]() {
      if (auto locked_node = weak_node.lock()) {
        locked_node->switch_off();
      }
    });
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: automatepro_python_tutorials/automatepro_python_tutorials/io/digital_out.py
import signal

import rclpy
from rclpy.duration import Duration
from rclpy.node import Node
from rclpy.signals import SignalHandlerOptions
from automatepro_interfaces.msg import DigitalOut


class DigitalOutPublisher(Node):

    def __init__(self):
        super().__init__('digital_out_publisher')
        self.publisher_ = self.create_publisher(DigitalOut, '/io/digital_out', 10)
        self.timer = self.create_timer(1.0, self.timer_callback)  # 1s
        self.duty_cycle_sequence = [0, 50, 100, 50]
        self.sequence_index = 0

    def timer_callback(self):
        msg = DigitalOut()
        msg.d_out_pin_id = DigitalOut.DIGITAL_OUT_H_01  # Digital Out Pin 01
        # Duty Cycle: 0%, 50%, 100%, 50%
        # 0% - OFF, 100% - ON
        msg.duty_cycle_percent = self.duty_cycle_sequence[self.sequence_index]
        self.publisher_.publish(msg)
        self.get_logger().info('Publishing: "%s"' % msg)
        self.sequence_index = (self.sequence_index + 1) % len(self.duty_cycle_sequence)

    def switch_off(self):
        msg = DigitalOut()
        msg.d_out_pin_id = DigitalOut.DIGITAL_OUT_H_01
        msg.duty_cycle_percent = 0
        self.publisher_.publish(msg)
        # Fast DDS acknowledges on the writer heartbeat, sent every 3 s by default.
        if self.publisher_.wait_for_all_acked(Duration(seconds=4)):
            self.get_logger().info('Switched DIGITAL_OUT_H_01 off')
        else:
            self.get_logger().warn('DIGITAL_OUT_H_01 off command not acknowledged')


def main(args=None):
    # The IO controller keeps the last command, so switch the output off before exiting.
    rclpy.init(args=args, signal_handler_options=SignalHandlerOptions.NO)
    signal.signal(signal.SIGTERM, signal.default_int_handler)
    node = DigitalOutPublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.switch_off()
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()
